BUG: Try to sort result of Index.union rather than guessing sortability

toobaz opened the pull request:
- [x] closes #17376
- [x] tests added / passed
- [x] passes `git diff master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The lines of code I removed seemed written with the intention of guaranteeing the same result under Python 2 and 3, but the docs say "Form the union of two Index objects and sorts _if possible._": so if it is possible in Python 2 we should do it, regardless of whether it is in Python 3. After all, this is how ``.sort_values()`` works. And more practically, trying to guess whether the result of an union can be sorted is hard.

toobaz committed:
BUG: Try to sort result of Index.union rather than guessing sortability

closes #17376

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -2195,35 +2195,21 @@ def union(self, other):
-                    warnings.warn("%s, sort order is undefined for "
-                                  "incomparable objects" % e, RuntimeWarning,
-                                  stacklevel=3)
+            try:
```
you probably can simply use ``core.sorting.safe_sort`` here instead and just remove the warning entirely (as it fails in many less cases)

toobaz replied:
Done, but I left the warning as it's still raised if not even ``safe_sort`` can sort them.

(I assume the failure is a problem with AppVeyor)

toobaz commented:
Any idea about what's happening with AppVeyor? It fails reliably on this PR with 2.7, but the trace doesn't point at any specific test.

gfyoung commented:
@toobaz : Nope, not a clue :cry: .  There's no way your changes AFAICT could have caused internal errors on the build.  I restarted AppVeyor to double check.

toobaz commented:
> I restarted AppVeyor to double check.

Failed again... shall I try closing this and opening a new PR?

jreback commented:
> Failed again... shall I try closing this and opening a new PR?

NO. that just creates more work. let me look.

jreback commented:
ok this is a legitimate catch. pytest-dist is segfaulting, hiding the traceback. So a couple of options to debug.

setup a vm on 3.6/windows is easiest / best.

you can temporarily change the options in the test runner (in ci/scripts_multiple.py) to something like:
    ``pytest -v -r xX -m "not single" --junitxml=/tmp/multiple.xml $TEST_ARGS pandas``

which disables the multi-processing and shows each test (so you can see where its failing).

jreback commented:
it *looks* like its either segfaulting or in a recursive loop somewhere.

toobaz commented:
> setup a vm on 3.6/windows is easiest / best.

Do you mean Python 3.6?! (tests fail only on 2.7)

gfyoung commented:
@toobaz  : I think @jreback probably meant 2.7 :smile:

jreback commented:
@toobaz this is related to #17839, could simply recast this to add the ``sort=`` kwarg (I think for back compat we have to use ``=False``, but maybe be able to change that, not sure).

toobaz commented:
> @toobaz this is related to #17839, could simply recast this to add the sort= kwarg (I think for back compat we have to use =False, but maybe be able to change that, not sure).

Certainly related, but I'm confused: are you suggesting the two methods should share more code? Or just that ``Index.union`` too should have a ``sort=`` argument? In this case I think that for backward compatibility it should default to ``True``...

jreback commented:
> Certainly related, but I'm confused: are you suggesting the two methods should share more code? Or just that Index.union too should have a sort= argument? In this case I think that for backward compatibility it should default to True...

yes I would add a ``sort=True`` kwarg.

jreback commented:
can you rebase. I think we do want to add this kw.

toobaz commented:
Unfortunately the problem on Windows + Python 2.7 persists. I haven't found the time to set up a virtual machine to debug.

jreback commented:
@toobaz if you can rebase would be great. I would also be ok with simply adding the keyword w/o changing *much* code as a first step.

jreback commented:
closing as stale. would like to generaly add the ``sort=`` kwarg though.

gfyoung commented:
@jreback : Let me take a crack at reviving this.

toobaz commented:
@gfyoung great! If you want a rebase, feel free to ask.

gfyoung commented:
There's a lint error on Travis, but I want to check the other builds before pushing the fix.

gfyoung commented:
@toobaz : That windows error looks puzzling.  Could be something with the build itself, as I can't see that "internal error" occurring locally at least.

TomAugspurger commented:
I don't think it conflicts much with this PR, but FYI I have a branch started adding `sort` to `union`. https://github.com/pandas-dev/pandas/compare/master...TomAugspurger:index-sort

I think it works for every index except `RangeIndex`, since some `RangeIndex.union(RangeIndex)` ops that return a RangeIndex with `sort=True` won't return a `RangeIndex` with `sort=False`.

gfyoung commented:
@TomAugspurger : Good to know.  Thanks!

gfyoung commented:
@toobaz : also not sure why you're running up against some random `clip` errors on Travis.  I can't reproduce those locally.

TomAugspurger commented:
I'm guessing it's a change to NumPy dev:
https://github.com/pandas-dev/pandas/issues/19976 (haven't tried locally
though)

On Fri, Mar 2, 2018 at 3:44 PM, gfyoung <notifications@github.com> wrote:

> @toobaz <https://github.com/toobaz> : also not sure why you're running up
> against some random clip errors on Travis. I can't reproduce those
> locally.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/17378#issuecomment-370061329>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIjAXOWBI_zEkcoeT0sOE4y4oscp3ks5tab0zgaJpZM4PG7KR>
> .
>

gfyoung commented:
> I'm guessing it's a change to NumPy dev:

@TomAugspurger : That seems reasonable.  I don't see any recent PR's for `clip`.

gfyoung committed:
CLN: Fix flake8 errors

gfyoung commented:
@toobaz : Despite the massive typo in my commit message, the latest commit is a dummy one to determine if that last AppVeyor build is just busted, or there is potentially an issue in the code.  I will undo that commit of course to restore the changes back in the PR.

toobaz commented:
> @toobaz : Despite the massive typo in my commit message, the latest commit is a dummy one to determine if that last AppVeyor build is just busted, or there is potentially an issue in the code

Fingers crossed... but I had triggered that behavior consistently (on multiple rebases).

gfyoung commented:
@toobaz : Not looking good here...the build passed just fine when I did all of your changes.

gfyoung commented:
So adding more verbosity to **pytest** appears to have given a hint as to the problematic tests:

https://ci.appveyor.com/project/pandas-dev/pandas/build/1.0.11213/job/36lok03ygfrk2g86#L5503

gfyoung commented:
Hmmm...so this issue might be known with pytest (or its plugins) already:

https://github.com/pytest-dev/pytest/issues/2223

Still not sure how @toobaz is triggering that though.

toobaz commented:
Maybe we could ask whether there is a way to use this [catchsegv](http://www.unix.com/man-page/linux/1/catchsegv/) within a pytest run?

gfyoung commented:
> Maybe we could ask whether there is a way to use this catchsegv within a pytest run?

@toobaz : Well, this is a Windows machine, not UNIX 😢 

I notice that we are using **pytest-forked**, which AFAICT, isn't well-supported on Windows:

https://github.com/pytest-dev/pytest-forked

I was about to file an issue against **pytest-xdist**, but I wonder if this might be the problem child.

toobaz commented:
> @toobaz : Well, this is a Windows machine, not UNIX cry

right 🙄

> I was about to file an issue against pytest-xdist, but I wonder if this might be the problem child.

What makes you think this is a bug in pytest, not a segfault somewhere in pandas? (or do you think it's both?)

gfyoung commented:
> What makes you think this is a bug in pytest, not a segfault somewhere in pandas? (or do you think it's both?)

I'm highly skeptical that your changes are causing a segfault.  That would be pretty impressive actually given that you're only touching high-level Python 😄

gfyoung committed:
[DO NOT MERGE] Is it pytest-xdist?

This will probably slow down performance,
but I can't see where the bug is in the changes.

gfyoung commented:
So removing xdist + rebasing somehow resulted in test failures on the first machine?

Going to leave this PR alone for now until `master` is patched,

toobaz commented:
> I'm highly skeptical that your changes are causing a segfault. That would be pretty impressive actually given that you're only touching high-level Python

Sure, that would require that my changes are triggering a bug somewhere in lower-level code... but I consider this more probable than a bug in pydist which affects precisely that test, in precisely this PR.

Then again, I'm definitely not an expert.

gfyoung commented:
I suppose, but if it's a bug in your code, I haven't been able to reproduce it locally.

toobaz commented:
> I suppose, but if it's a bug in your code, I haven't been able to reproduce it locally.

Aha! I had missed that you were testing on Windows. Good to know.

jreback commented on appveyor.yml:
```diff
@@ -73,7 +73,7 @@ install:
 
   # create our env
-  - cmd: conda create -n pandas python=%PYTHON_VERSION% cython pytest>=3.1.0 pytest-xdist
+  - cmd: conda create -n pandas python=%PYTHON_VERSION% cython pytest>=3.1.0
```
so does xdist cause the issues with this?

gfyoung commented:
@toobaz @jreback : So good news (I guess?), I was able to replicate the AppVeyor failure locally.

gfyoung commented:
@toobaz : It appears that you loosened the restrictions too much, at least for Windows:

* The test that is breaking AppVeyor is **TestAlignment.test_basic_frame_alignment**
* The condition where **len(indexer) > 0** is the problem.  For some reason, I need to put back the comparison of `lvals[0]` and `other_diff[0]` to remove the break OR only trying sorting when in the `else` branch (i.e. when `len(indexer) = 0`).

toobaz commented:
@gfyoung thanks for your analysis!

Does the problem arise with any specific combination of index types? (from ``args = product(...``)

gfyoung commented:
> Does the problem arise with any specific combination of index types?

I can't tell unfortunately.  It seems to have crashed with multiple types.

jreback commented:
@toobaz whats' the status on this?

jreback commented:
closing as stale, if you want to continue working, pls ping and we can re-open. you will need to merge master.
